.github/workflows/ci_tests.yaml

```diff
@@ -160,8 +160,10 @@ jobs:
 
       # Upload coverage to Codecov
       - name: Upload coverage to Codecov
-        uses: codecov/codecov-action@v3.1.4
+        uses: codecov/codecov-action@v4.0.1
         with:
           file: ./coverage.xml # optional
           env_vars: OS,PYTHON,NUMPY
           fail_ci_if_error: false
+        env:
+          CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
```

Comment:
Note that [codecov/codecov-action v4](https://github.com/codecov/codecov-action/releases/tag/v4.0.0) requires a CODECOV_TOKEN to be set (tokenless uploads are no longer supported). Looks like that token is already available as a secret in our CI (probably from #2792), so I just passed it in as an environment variable here.